contracts/test/v0.8/PermissionedForwardProxy.ts

```diff
@@ -0,0 +1,168 @@
+import { ethers } from 'hardhat'
+import { publicAbi } from '../test-helpers/helpers'
+import { expect } from 'chai'
+import { Contract, ContractFactory } from 'ethers'
+import { Personas, getUsers } from '../test-helpers/setup'
+
+let personas: Personas
+
+let controllerFactory: ContractFactory
+let counterFactory: ContractFactory
+let controller: Contract
+
+before(async () => {
+  personas = (await getUsers()).personas
+  controllerFactory = await ethers.getContractFactory(
+    'src/v0.8/PermissionedForwardProxy.sol:PermissionedForwardProxy',
+    personas.Carol,
+  )
+  counterFactory = await ethers.getContractFactory(
+    'src/v0.8/tests/Counter.sol:Counter',
+    personas.Carol,
+  )
+})
+
+describe('PermissionedForwardProxy', () => {
+  beforeEach(async () => {
+    controller = await controllerFactory.connect(personas.Carol).deploy()
+  })
+
+  it('has a limited public interface [ @skip-coverage ]', async () => {
+    publicAbi(controller, [
+      'forward',
+      'addPermission',
+      'removePermission',
+      'forwardPermissionList',
+      // Owned
+      'acceptOwnership',
+      'owner',
+      'transferOwnership',
+    ])
+  })
+
+  describe('#addPermission', () => {
+    describe('when called by a non-owner', () => {
+      it('reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Eddy)
+            .addPermission(
+              await personas.Carol.getAddress(),
+              await personas.Eddy.getAddress(),
+            ),
+        ).to.be.revertedWith('Only callable by owner')
+      })
+    })
+
+    describe('when called by the owner', () => {
+      beforeEach(async () => {
+        // Reset permission before testing
+        await controller
```

Comment:
shouldn't have to do this if you deploy a new controller before each test, right?

Reply:
(y) i wasn't really sure how before each works with recursive describes. Will try to remove this

Reply:
a beforeEach in a parent `describe()` will still run before tests inside of child `describe()` locks